hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java

```diff
@@ -1558,13 +1558,15 @@ private void tryUpgrade(HoodieTableMetaClient metaClient, Option<String> instant
         new UpgradeDowngrade(metaClient, config, context, upgradeDowngradeHelper);
 
     if (upgradeDowngrade.needsUpgradeOrDowngrade(HoodieTableVersion.current())) {
-      // Ensure no inflight commits by setting EAGER policy and explicitly cleaning all failed commits
-      List<String> instantsToRollback = getInstantsToRollback(metaClient, HoodieFailedWritesCleaningPolicy.EAGER, instantTime);
+      if (config.getWriteConcurrencyMode().supportsOptimisticConcurrencyControl()) {
+        // Ensure no inflight commits by setting EAGER policy and explicitly cleaning all failed commits
+        List<String> instantsToRollback = getInstantsToRollback(metaClient, HoodieFailedWritesCleaningPolicy.EAGER, instantTime);
 
-      Map<String, Option<HoodiePendingRollbackInfo>> pendingRollbacks = getPendingRollbackInfos(metaClient);
-      instantsToRollback.forEach(entry -> pendingRollbacks.putIfAbsent(entry, Option.empty()));
+        Map<String, Option<HoodiePendingRollbackInfo>> pendingRollbacks = getPendingRollbackInfos(metaClient);
+        instantsToRollback.forEach(entry -> pendingRollbacks.putIfAbsent(entry, Option.empty()));
```

Comment:
The pre-upgrade rollback was introduced in https://github.com/apache/hudi/pull/2374, with a pre-condition that when the write concurrency mode is based on optimistic lock.

While in https://github.com/apache/hudi/pull/4114, the pre-condition was removed, which i think is a regression because we already do the rollback when starting new instant/commit before write. In single writer mode, there is no need to do redundant rollback again before upgrade.

Reply:
@n3nash : change is ok with me. Do you see any issues as such. 
CC @alexeykudinkin

Reply:
One case that I can think of is when the CLEAN_FAILED_FILES policy is set to LAZY and writer config is set to a single writer - I believe there can be older inflight files lying around. In this case if the rollback is done only for Optimistic Concurrency, some older inflight metadata files may be lying around.

Reply:
> some older inflight metadata files may be lying around

That's true, but what is the problem here in single writer upgrade ? It seems that the problem is not about upgrade but about the metadata table rollback, but shouldn't the metadata table also has MVCC/SI transaction semantics here ? The metadata table reader can not see these inflight meta logs right ?

Reply:
@danny0405 I'm not fully following your comment. From this PR, what I gather is that we want to special case rollback for Optimistic Concurrency only since for Single Writer this will be redundant given rollback is supposed to happen at the start of the commit. I'm merely saying that's not necessarily the case due to the LAZY cleaning policy, does that make sense ?

Reply:
hey @danny0405 : I am confused as to whats the regression here. I thought this is more of an optimization. From your first comment "While in https://github.com/apache/hudi/pull/4114, the pre-condition was removed, which i think is a regression because we already do the rollback when starting new instant/commit before write. In single writer mode, there is no need to do redundant rollback again before upgrade."

can you help me understand why would be do redundant rollback. once we rolledback a given commit, next time if we check if something needs to be rolled back, the same partially failed commit should not appear right. I mean, if the rollback was complete the first time, it should be be part of the list of commits to rollback the 2nd time right. or am I missing something here.

Reply:
Hi @nsivabalan , i agree with your point, but at least, there is no need for redundant metadata checking if the rollback strategy is EAGER.

And still, no one answers the question why we need the rollback (for data set files and metadata files) before upgrade ? What is exactly the purpose here ? 

@n3nash  say that it is because we need to rollback the metadata table if it is enabled, but we can not see any clues that the rollback is only for metadata. And even if we do not rollback the metadata, the content read from the metadata should still be correct based on that the metadata table itself is a Hudi table and has SI visibility ?

Seems it is clear that the rollback here is not for data set files, correct me if i'm wrong.

Reply:
Hello @nsivabalan @n3nash , i have refactored the code a little to

1. move the failed writes cleaning all together to when starting the new commit
2. the cleaning strategy for commit actions as tweaked: force to clean if metadata table is enabled in case we commit metadata table successfully but data set table failed

Please review again if you have time, thanks so much in advance ~

Reply:
I analyze the code a little and find that we would do the metadata table sync when it is initialized, so for the case "commit metadata successfully but data set table failed", it should be fine.

Also i didn't find any upgrade/downgrade code that needs the data set/metadata files, from my perspective, the rollback is totally unnecessary, so i just remove it here.